pkg/etcd/etcd.go

```diff
@@ -1048,18 +1049,59 @@ func (e *ETCD) manageLearners(ctx context.Context) {
 			continue
 		}
 
+		nodes, err := e.parseNodesWithEtcd()
+		if err != nil {
+			logrus.Errorf("Error while listing nodes with etcd status: %v", err)
+			continue
+		}
+
 		for _, member := range members.Members {
 			if member.IsLearner {
 				if err := e.trackLearnerProgress(ctx, progress, member); err != nil {
 					logrus.Errorf("Failed to track learner progress towards promotion: %v", err)
 				}
+
+				// If the node exists in the parse and is a learner, then the status condition is set.
+				if node, ok := nodes[member.Name]; ok {
+					_, err := e.setEtcdStatusCondition(&node, false)
+					if err != nil {
+						logrus.Errorf("Unable to set etcd status condition %s: %v", member.Name, err)
+					}
+				}
+
 				break
 			}
+
+			// If the node exists in the parse and is a voter, then the status condition is set.
+			if node, ok := nodes[member.Name]; ok {
+				_, err := e.setEtcdStatusCondition(&node, true)
+				if err != nil {
+					logrus.Errorf("Unable to set etcd status condition %s: %v", member.Name, err)
+				}
+			}
 		}
 	}
 	return
 }
 
+// parseNodesWithEtcd searches for nodes that have an annotation with etcd.{version.Program}.cattle.io/node-name
+// and sets a hashMap for easy node retrieval by node name.
+func (e *ETCD) parseNodesWithEtcd() (map[string]v1.Node, error) {
+	newMap := map[string]v1.Node{}
+	nodes, err := e.config.Runtime.Core.Core().V1().Node().List(metav1.ListOptions{})
```

Comment:
The correct way to do this would be to use a labelSelector and filter by the etcd role label.
https://github.com/k3s-io/k3s/blob/49411e7084f74b931882f1cf85bf2e17a5f02158/pkg/etcd/snapshot_controller.go#L178-L183